Replace hand-rolled SharedMutex with C++17 std::shared_mutex (NFCI)

adrian-prantl opened the pull request:
We are getting reports where LLDB is hanging in try_lock().  This patch makes the code easier to reason about by using standard primitives, and avoids the pitfall of the copy constructor that the old code was still allowing. While this may not fix the hangs, it should certainly make it easier to diagnose.

rdar://112025620

adrian-prantl committed:
Replace hand-rolled SharedMutex with C++17 std::shared_mutex (NFCI)

We are getting reports where LLDB is hanging in try_lock().  This
patch makes the code easier to reason about by using standard
primitives, and avoids the pitfall of the copy constructor that the
old code was still allowing. While this may not fix the hangs, it
should certainly make it easier to diagnose.

rdar://112025620

adrian-prantl commented:
@swift-ci test

JDevlieghere commented:
LLVM's `RWMutex` uses `shared_mutex` if the deployment target allows it. Either this is a NOOP or it has the potential to break bots with a lower deployment target.

adrian-prantl commented:
> LLVM's `RWMutex` uses `shared_mutex` if the deployment target allows it. Either this is a NOOP or it has the potential to break bots with a lower deployment target.

According to the comment in RWMutex.h std::shared_timed_mutex is only available on macOS 10.12 and later. Green dragon is running on 10.15.5.

Note that `llvm::RWMutex` does not have a `try_lock` method and it's my hand-rolled implementation of that, that I would like to remove with this patch.

JDevlieghere commented:
> According to the comment in RWMutex.h std::shared_timed_mutex is only available on macOS 10.12 and later. Green dragon is running on 10.15.5.

Guess who wrote that comment. 🙃 Also, I assume you mean Swift CI? If both GreenDragon and Swift CI are using 10.15 as the minimum deployment target I'll remove that workaround upstream. 

> Note that `llvm::RWMutex` does not have a `try_lock` method and it's my hand-rolled implementation of that, that I would like to remove with this patch.

👍

adrian-prantl commented:
ci.swift.org claims to use `Host OS: macOS 12.6`.

adrian-prantl commented:
I just realized that the "host os" != "deployment target". Green dragon uses `x86_64-apple-darwin20.6.0` (macOS 11) and ci.swift.org seems to build for  `x86_64-apple-macosx10.13`.
So we should be good here.

JDevlieghere commented:
Actually, my memory was correct, but it seems someone hoisted `shared_mutex` out of the availability check. Seems like Google is still running a macOS bot that targets 10.12 (discussed in https://reviews.llvm.org/D130689) so let's keep the workaround in place for `shared_timed_mutex` upstream. That doesn't affect this patch.

kastiglione commented on lldb/source/Target/Target.cpp:
can this use `std::lock_guard` + `std::adopt_lock`?

```suggestion
          std::lock_guard<std::shared_mutex> unlock(lock, std::adopt_lock);
```

adrian-prantl replied:
Nice!

kastiglione commented on lldb/include/lldb/Core/SwiftScratchContextReader.h:
```diff
@@ -97,27 +47,25 @@ struct ScopedSharedMutexReader {
 /// An RAII object that just acquires the reader lock.
-struct SwiftScratchContextLock : ScopedSharedMutexReader {
+struct SwiftScratchContextLock {
+  std::shared_lock<std::shared_mutex> lock;
```
maybe `m_lock` for consistency?

adrian-prantl replied:
IIUC the lldb coding style uses `m_` only in classes, not in structs.

kastiglione replied:
TIL

adrian-prantl committed:
Hoist check outside of lambda (NFC)

adrian-prantl committed:
Fix a race condition.

The reader lock must be acquired before retrieving a scratch context,
otherwise another thread could still tear it down between the time it
was created and the creation of the reader object.

There are other shortcomings in GetSwiftScratchContext() that are
documented, but not addressed, in this patch to make it easier to review.

adrian-prantl committed:
Use std::adopt_lock instead of lambda

adrian-prantl commented:
@kastiglione @JDevlieghere I added another commit on top that actually fixes a concurrency issue.

adrian-prantl commented:
@swift-ci test

augusto2112 commented on lldb/source/Target/Target.cpp:
```diff
@@ -2716,62 +2716,69 @@ llvm::Optional<SwiftScratchContextReader> Target::GetSwiftScratchContext(
     if (!type_system_or_err) {
       llvm::consumeError(type_system_or_err.takeError());
-      return nullptr;
+      return;
```
Nit: we could log this error instead of silently consuming it.

augusto2112 commented on lldb/source/Target/Target.cpp:
```diff
@@ -2789,42 +2792,55 @@ llvm::Optional<SwiftScratchContextReader> Target::GetSwiftScratchContext(
-      log->Printf("returned project-wide scratch context\n");
+  llvm::Optional<SwiftScratchContextReader> reader;
+  if (lldb_module && m_use_scratch_typesystem_per_module) {
+    maybe_create_fallback_context();
```
Could you explain one more time why this can't potentially delete the scratch context while it's being used?

In my mind the following events could happen:

- Thread 1 successfully calls `GetSwiftScratchContext` and creates a new scratch type system.
- Thread 2 calls `GetSwiftScratchContext`
  - Thread 2 calls `maybe_create_fallback_context` (this call is not guarded by any locks as far as I can tell).
    - `maybe_create_fallback_context` deletes the scratch type system.
- Thread 1 is now using a dangling reference to the scratch type system.

adrian-prantl replied:
You are right.

adrian-prantl replied:
I addressed this by only deleting when we can acquire the write lock.

JDevlieghere commented on lldb/source/Target/Target.cpp:
```diff
@@ -2716,62 +2716,69 @@ llvm::Optional<SwiftScratchContextReader> Target::GetSwiftScratchContext(
-        return nullptr;
+        // FIXME: Shouldn't this continue rather than return?
+        m_scratch_typesystem_for_module.erase(key);
+        if (log)
+          log->Printf("erased module-wide scratch context with errors\n");
```
`LLDB_LOG`

JDevlieghere commented on lldb/source/Target/Target.cpp:
```diff
@@ -2785,41 +2792,52 @@ llvm::Optional<SwiftScratchContextReader> Target::GetSwiftScratchContext(
+  llvm::Optional<SwiftScratchContextReader> reader;
+  if (lldb_module && m_use_scratch_typesystem_per_module) {
+    maybe_create_fallback_context();
+    std::shared_lock<std::shared_mutex> lock(GetSwiftScratchContextLock());
```
Would it be possible to create a reader object from the get_or_create object and make the whole operation atomic?

adrian-prantl replied:
Unfortunately, no. AFAIK, there is no way to "downgrade" a unique lock into a shared lock, so we need to reacquire it. That means we also need to do the cache lookup again, because another thread could have swapped it out in the mean time by calling the same function.

JDevlieghere commented on lldb/source/Target/Target.cpp:
TIL!

augusto2112 commented on lldb/include/lldb/Core/SwiftScratchContextReader.h:
```diff
@@ -97,27 +47,25 @@ struct ScopedSharedMutexReader {
-class SwiftScratchContextReader : ScopedSharedMutexReader {
-  TypeSystemSwiftTypeRefForExpressions *m_ptr;
+class SwiftScratchContextReader {
+  std::shared_lock<std::shared_mutex> m_lock;
```
All the locks of the shared mutex are locked by `std::shared_lock` (none of them use `std::unique_lock`). Doesn't this mean that all of them can use the type system concurrently freely?

adrian-prantl replied:
Correct. The lock is not meant to ensure exclusive access, but guard against a context from being destroyed while still being used.

JDevlieghere replied:
Fair enough, you would have to release the unique_lock and acquire the shared_lock, which has the same drawback as the current approach.

JDevlieghere reviewed (approved):
LGTM

adrian-prantl commented:
@swift-ci test windows

adrian-prantl committed:
Only delete the scratch context if we can acquire the unique lock

adrian-prantl commented:
@swift-ci test

adrian-prantl commented:
@swift-ci test windows
